File: server/csharp/MetaObjects/Template/TemplateSchema.cs
// Template attribute schemas — the per-subtype attr sets for template.prompt,
// template.output, and template.toolcall.
//
// Colocated per ADR-0003. Mirrors the canonical spec/metamodel/template.json (which
// this port embeds + reads): DESCRIPTIONS are deliberately NOT hand-copied here — FR-033
// sources every description from the shared JSON via Registry.ApplySpecDescriptions, so
// the prose is byte-identical to TS by construction. Only the facets the manifest needs
// but the description pass does not carry (value type / array-ness / requiredness /
// allowedValues / default) are declared here.
//
// These attrs are OWN to the template type — a template.prompt with no @payloadRef
// can't render, a template.toolcall with no @toolName can't be wired to an LLM — so
// they register HERE, in the core provider (CoreTypes.cs), never in a concern
// provider that can be composed out. A required attr that lived in an optional
// provider would silently stop being enforced the moment that provider was dropped
// (the defect this schema fixes; see CoreTypes.cs "template" section). Mirrors
// RequirementSchema, the correct pattern for an own-attr type in this port.
//
// template.base carries no attrs at all — it is not in TemplateAttrsMap, so a
// lookup miss falls back to an empty list (mirrors RequirementAttrsMap, which has
// no "base" entry either).

using MetaObjects.Core.Attr;

namespace MetaObjects.Template;

/// <summary>Attribute schemas for the template concern.</summary>
public static class TemplateSchema
{
    // --- shared across prompt + output (some also shared with toolcall) ---

    /// <summary>@payloadRef — required on every concrete template subtype (prompt/output/toolcall).</summary>
    private static readonly AttrSchema PayloadRefAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_PAYLOAD_REF,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: true,
        Description: "Reference to the payload (a view-object / projection) this template renders against.");

    /// <summary>@textRef — prompt + output only (toolcall has no renderable text body).</summary>
    private static readonly AttrSchema TextRefAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_TEXT_REF,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: false,
        Description: "2-layer logical reference (group/source) to the body text, resolved by a provider at render time.");

    /// <summary>@format — prompt + output only; closed enum keyed by the render engine's escaper.</summary>
    private static readonly AttrSchema FormatAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_FORMAT,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: false,
        Default: TemplateConstants.TEMPLATE_FORMAT_DEFAULT,
        AllowedValues: [.. TemplateConstants.TEMPLATE_FORMATS],
        Description: "Output format; drives the render engine's escaping/whitespace behavior.");

    /// <summary>@maxChars — prompt + output only.</summary>
    private static readonly AttrSchema MaxCharsAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_MAX_CHARS,
        ValueType: AttrConstants.ATTR_SUBTYPE_INT,
        Required: false,
        Description: "Size budget for the rendered output, in characters.");

    /// <summary>@owner — governance; every concrete template subtype.</summary>
    private static readonly AttrSchema OwnerAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_OWNER,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: false,
        Description: "Governance: the owner of this template.");

    /// <summary>@since — governance; every concrete template subtype.</summary>
    private static readonly AttrSchema SinceAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_SINCE,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: false,
        Description: "Governance: the version this template was introduced in.");

    /// <summary>@requiredTags — prompt + output only.</summary>
    private static readonly AttrSchema RequiredTagsAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_REQUIRED_TAGS,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: false,
        IsArray: true,
        Description: "Output tags the rendered text must contain (drives the verify output-tag check).");

    /// <summary>@maxTokens — prompt + toolcall (peer LLM-specific budgets).</summary>
    private static readonly AttrSchema MaxTokensAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_MAX_TOKENS,
        ValueType: AttrConstants.ATTR_SUBTYPE_INT,
        Required: false,
        Description: "Token budget for the rendered prompt (LLM-specific).");

    // --- prompt-only overlay ---

    private static readonly AttrSchema RequiredSlotsAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_REQUIRED_SLOTS,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: false,
        IsArray: true,
        Description: "Slots that must resolve at render time (drives the verify check).");

    private static readonly AttrSchema ModelAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_MODEL,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: false,
        Description: "Target model id (LLM-specific).");

    private static readonly AttrSchema ResponseRefAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_RESPONSE_REF,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: false,
        Description: "Optional ref to the response value-object this prompt expects (peer of @payloadRef; drives typed LLM-call trace derivation).");

    // --- output-only overlay ---

    private static readonly AttrSchema PromptStyleAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_PROMPT_STYLE,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: false,
        Default: TemplateConstants.PROMPT_STYLE_DEFAULT,
        AllowedValues: [.. TemplateConstants.TEMPLATE_PROMPT_STYLES],
        Description: "FR-010 output-format prompt presentation: 'guide' (prose list + example), 'inline' (inline placeholders / enum choices), or 'exampleOnly' (filled skeleton). Guidance is never emitted as comments.");

    private static readonly AttrSchema KindAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_KIND,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: false,
        Default: TemplateConstants.TEMPLATE_KIND_DEFAULT,
        AllowedValues: [.. TemplateConstants.TEMPLATE_KINDS],
        Description: "Output shape: 'document' (renders @textRef in @format → one string) or 'email' (renders subject + html + optional text → a structured EmailDocument).");

    private static readonly AttrSchema SubjectRefAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_SUBJECT_REF,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: false,
        Description: "Email only: 2-layer logical reference (group/source) to the subject-line text. Required when @kind=\"email\".");

    private static readonly AttrSchema HtmlBodyRefAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_HTML_BODY_REF,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: false,
        Description: "Email only: 2-layer logical reference (group/source) to the HTML body text. Required when @kind=\"email\".");

    private static readonly AttrSchema TextBodyRefAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_TEXT_BODY_REF,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: false,
        Description: "Email only: 2-layer logical reference (group/source) to the optional plain-text alternative body.");

    // --- toolcall-only overlay ---

    private static readonly AttrSchema ToolNameAttr = new AttrSchema(
        Name: TemplateConstants.TEMPLATE_ATTR_TOOL_NAME,
        ValueType: AttrConstants.ATTR_SUBTYPE_STRING,
        Required: true,
        Description: "Wire tool name surfaced to the LLM (vendor-specific format).");

    // --- per-subtype attr lists (JSON declaration order; the manifest emitter sorts) ---

    private static readonly IReadOnlyList<AttrSchema> PromptAttrs =
    [
        PayloadRefAttr,
        TextRefAttr,
        FormatAttr,
        MaxCharsAttr,
        OwnerAttr,
        SinceAttr,
        RequiredTagsAttr,
        MaxTokensAttr,
        RequiredSlotsAttr,
        ModelAttr,
        ResponseRefAttr,
    ];

    private static readonly IReadOnlyList<AttrSchema> OutputAttrs =
    [
        PayloadRefAttr,
        TextRefAttr,
        FormatAttr,
        MaxCharsAttr,
        OwnerAttr,
        SinceAttr,
        RequiredTagsAttr,
        PromptStyleAttr,
        KindAttr,
        SubjectRefAttr,
        HtmlBodyRefAttr,
        TextBodyRefAttr,
    ];

    private static readonly IReadOnlyList<AttrSchema> ToolcallAttrs =
    [
        ToolNameAttr,
        PayloadRefAttr,
        OwnerAttr,
        SinceAttr,
        MaxTokensAttr,
    ];

    /// <summary>
    /// Attrs per template subtype. prompt has 11, output has 12, toolcall has 5
    /// (matches spec/metamodel/template.json). template.base carries none — it is
    /// deliberately absent from this map; a lookup miss is the caller's cue to fall
    /// back to an empty list (mirrors RequirementAttrsMap, which has no "base" entry).
    /// </summary>
    public static readonly IReadOnlyDictionary<string, IReadOnlyList<AttrSchema>> TemplateAttrsMap =
        new Dictionary<string, IReadOnlyList<AttrSchema>>
        {
            [TemplateConstants.TEMPLATE_SUBTYPE_PROMPT] = PromptAttrs,
            [TemplateConstants.TEMPLATE_SUBTYPE_OUTPUT] = OutputAttrs,
            [TemplateConstants.TEMPLATE_SUBTYPE_TOOLCALL] = ToolcallAttrs,
        };
}

File: server/typescript/packages/metadata/src/template/template-schema.ts
// Reserved-attribute schemas per template subtype (FR-004, R1).
//
// Both subtypes share the generic attrs; @format is a closed enum (allowedValues)
// that the render engine keys its escaper off. template.prompt additionally
// carries the LLM-overlay attrs. References (@payloadRef, @textRef) are plain
// strings here — resolution against a real payload / text source is render-time
// `verify` scope (Plan #3), not load-time validation.

import type { AttrSchema } from "../registry.js";
import {
  ATTR_SUBTYPE_STRING,
  ATTR_SUBTYPE_INT,
  ATTR_SUBTYPE_STRINGARRAY,
} from "../core/attr/attr-constants.js";
import { SUBTYPE_BASE } from "../shared/base-types.js";
import {
  TEMPLATE_SUBTYPE_PROMPT,
  TEMPLATE_SUBTYPE_OUTPUT,
  TEMPLATE_SUBTYPE_TOOLCALL,
  TEMPLATE_ATTR_PAYLOAD_REF,
  TEMPLATE_ATTR_TEXT_REF,
  TEMPLATE_ATTR_FORMAT,
  TEMPLATE_ATTR_MAX_CHARS,
  TEMPLATE_ATTR_OWNER,
  TEMPLATE_ATTR_SINCE,
  TEMPLATE_ATTR_REQUIRED_TAGS,
  TEMPLATE_ATTR_MAX_TOKENS,
  TEMPLATE_ATTR_REQUIRED_SLOTS,
  TEMPLATE_ATTR_MODEL,
  TEMPLATE_ATTR_TOOL_NAME,
  TEMPLATE_FORMATS,
} from "./template-constants.js";

// Generic attrs shared by template.prompt and template.output.
const genericAttrs: AttrSchema[] = [
  {
    name: TEMPLATE_ATTR_PAYLOAD_REF,
    valueType: ATTR_SUBTYPE_STRING,
    required: true,
    description: "Reference to the payload (a view-object / projection) this template renders against.",
  },
  {
    name: TEMPLATE_ATTR_TEXT_REF,
    valueType: ATTR_SUBTYPE_STRING,
    required: true,
    description: "2-layer logical reference (group/source) to the body text, resolved by a provider at render time.",
  },
  {
    name: TEMPLATE_ATTR_FORMAT,
    valueType: ATTR_SUBTYPE_STRING,
    required: false,
    default: "text",
    allowedValues: [...TEMPLATE_FORMATS],
    description: "Output format; drives the render engine's escaping/whitespace behavior.",
  },
  {
    name: TEMPLATE_ATTR_MAX_CHARS,
    valueType: ATTR_SUBTYPE_INT,
    required: false,
    description: "Size budget for the rendered output, in characters.",
  },
  {
    name: TEMPLATE_ATTR_OWNER,
    valueType: ATTR_SUBTYPE_STRING,
    required: false,
    description: "Governance: the owner of this template.",
  },
  {
    name: TEMPLATE_ATTR_SINCE,
    valueType: ATTR_SUBTYPE_STRING,
    required: false,
    description: "Governance: the version this template was introduced in.",
  },
  {
    name: TEMPLATE_ATTR_REQUIRED_TAGS,
    valueType: ATTR_SUBTYPE_STRINGARRAY,
    required: false,
    description: "Output tags the rendered text must contain (drives the verify output-tag check).",
  },
];

// LLM-overlay attrs (template.prompt only).
const promptOverlayAttrs: AttrSchema[] = [
  {
    name: TEMPLATE_ATTR_MAX_TOKENS,
    valueType: ATTR_SUBTYPE_INT,
    required: false,
    description: "Token budget for the rendered prompt (LLM-specific).",
  },
  {
    name: TEMPLATE_ATTR_REQUIRED_SLOTS,
    valueType: ATTR_SUBTYPE_STRINGARRAY,
    required: false,
    description: "Slots that must resolve at render time (drives the verify check).",
  },
  {
    name: TEMPLATE_ATTR_MODEL,
    valueType: ATTR_SUBTYPE_STRING,
    required: false,
    description: "Target model id (LLM-specific).",
  },
];

// Toolcall attrs (template.toolcall only — does NOT inherit genericAttrs).
// Per ADR-0011: vendor-agnostic in core; vendor wire details (retry semantics,
// fallback shapes, parallel invocation, cache hints) added by consumer
// providers via registry.extend(TYPE_TEMPLATE, "toolcall", { attributes: [...] }).
//
// Critical: @textRef is intentionally NOT required here. A tool-call has no
// renderable text body — the body IS the structured output schema resolved
// via @payloadRef. This is the design rationale for toolcall being its own
// subtype rather than template.output + @toolName.
//
// @description is intentionally NOT declared here — it's already a documentation
// common attr added to every type by docProvider. Tool descriptions surfaced to
// the LLM read the same @description common attr that doc-gen uses.
const toolcallAttrs: AttrSchema[] = [
  {
    name: TEMPLATE_ATTR_TOOL_NAME,
    valueType: ATTR_SUBTYPE_STRING,
    required: true,
    description: "Wire tool name surfaced to the LLM (vendor-specific format).",
  },
  {
    name: TEMPLATE_ATTR_PAYLOAD_REF,
    valueType: ATTR_SUBTYPE_STRING,
    required: true,
    description: "Output value-object the tool produces (resolved against the metamodel).",
  },
  {
    name: TEMPLATE_ATTR_OWNER,
    valueType: ATTR_SUBTYPE_STRING,
    required: false,
    description: "Governance: the owner of this toolcall.",
  },
  {
    name: TEMPLATE_ATTR_SINCE,
    valueType: ATTR_SUBTYPE_STRING,
    required: false,
    description: "Governance: the version this toolcall was introduced in.",
  },
];

export const TEMPLATE_ATTRS_MAP = new Map<string, AttrSchema[]>([
  [SUBTYPE_BASE, []],
  [TEMPLATE_SUBTYPE_PROMPT, [...genericAttrs, ...promptOverlayAttrs]],
  [TEMPLATE_SUBTYPE_OUTPUT, [...genericAttrs]],
  [TEMPLATE_SUBTYPE_TOOLCALL, [...toolcallAttrs]],
]);
